### PR TITLE
Calculate average L1 gas usage statistics

### PR DESCRIPTION
**Pull request by andrewgordstewart**
**Depends on https://github.com/statechannels/SAFE-protocol/pull/84**

Includes the following changes:
- all deps are devDeps
- a `printScenarioGasUsage` prints a nice table (see example below)
- test helpers are refactored to allow a dynamic number of tickets to be executed in a `swap` call
- L1 gas benchmarking is done for `1,2,5,20,50,86` tickets per batch

### Example output
Screen Shot 2021-12-16 at 1 32 00 PM

### CODE REVIEW

File: test/safe.test.ts [hunk 1]
```diff
@@ -56,34 +58,59 @@ async function deposit(trustedNonce: number, trustedAmount: number) {
...
-async function swap(trustedNonce: number, trustedAmount: number) {
-  await deposit(trustedNonce, trustedAmount);
-  const [ticket, ticket2, signature] = await authorizeWithdrawal(trustedNonce);
-
-  await waitForTx(lpL1.claimBatch([ticket, ticket2], signature));
-
-  await ethers.provider.send("evm_increaseTime", [121]);
+/**
+ * 
+ * @param trustedNonce ???
```
Comment (kerzhner) [inline test/safe.test.ts:95]:
```suggestion
 * @param trustedNonce The sum of all tickets starting with trustedNonce + new deposit must be <= trustedAmount
```

File: test/safe.test.ts [hunk 2]
```diff
@@ -56,34 +58,59 @@ async function deposit(trustedNonce: number, trustedAmount: number) {
...
+  const { tickets, signature } = await authorizeWithdrawal(trustedNonce, numTickets);
+
+  const l1TransactionReceipt = await waitForTx(lpL1.claimBatch(tickets, signature, { gasLimit }));
+
+  await ethers.provider.send("evm_increaseTime", [SAFETY_DELAY + 1]);
   await waitForTx(lpL2.claimL2Funds(trustedNonce));
+
+  // TODO: This ought to estimate the total user cost. The cost of the L1 transaction
+  // is currently used as a rough estimate of the total user cost.
+  return l1TransactionReceipt
```
Comment (kerzhner) [inline test/safe.test.ts:113]:
```suggestion
  return l1TransactionReceipt;
```

File: test/utils.ts [hunk 3]
```diff
@@ -1,21 +1,33 @@
-import { BigNumber, Contract, Wallet } from "ethers";
-import { USE_ERC20 } from "../src/constants";
-import { IERC20 } from "../contract-types/IERC20";
+import Table from "cli-table";
+import { BigNumber } from "ethers";
 
-export type Balances = { alice: BigNumber; bob: BigNumber };
+export type ClaimTicketsScenario = {
```
Comment (kerzhner) [inline test/utils.ts:4]: Doesn't seem like this needs to be exported.

File: yarn.lock [hunk 4]
```diff
@@ -10546,11 +10571,11 @@ __metadata:
 
 "typescript@patch:typescript@^4.5.4#~builtin<compat/typescript>":
   version: 4.5.4
-  resolution: "typescript@patch:typescript@npm%3A4.5.4#~builtin<compat/typescript>::version=4.5.4&hash=ddd1e8"
+  resolution: "typescript@patch:typescript@npm%3A4.5.4#~builtin<compat/typescript>::version=4.5.4&hash=493e53"
```
Comment (kerzhner) [inline yarn.lock:10574]: The dreaded typescript hash change.

**Review by kerzhner — APPROVED** [review]

File: test/utils.ts (hunk 3 above)
Comment (andrewgordstewart) [reply test/utils.ts]: 3d5d8cbe01209f4668bc961e76bb03543d95ba3e

**Pull request merged**